src/xml_parser_main_metadata_together.c

```diff
@@ -337,133 +351,215 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
                                         gboolean allow_out_of_order,
                                         GError **err)
 {
-    int ret = CRE_OK;
-    CR_FILE *primary_f = NULL;
-    CR_FILE *filelists_f = NULL;
-    CR_FILE *other_f = NULL;
-    GError *tmp_err = NULL;
+    assert(pkgcb || newpkgcb);
+    cr_PkgIterator* pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, newpkgcb, newpkgcb_data, warningcb, warningcb_data, allow_out_of_order, err
+    );
 
-    cr_CbData cbdata;
-    cbdata.in_progress_pkgs_list = NULL;
-    cbdata.in_progress_pkgs_hash = NULL;
-    cbdata.newpkgcb = newpkgcb;
-    cbdata.newpkgcb_data = newpkgcb_data;
-    cbdata.pkgcb = pkgcb;
-    cbdata.pkgcb_data = pkgcb_data;
+    // TODO: do this more properly
+    assert(pkg_iterator);
+    assert(!*err);
```

Comment:
I would prefer if instead of the `assert(!*err);` we return the `err` code, perhaps:
```suggestion
    if (*err) {
        return (*err)->code;
    }
    assert(pkg_iterator);
```
I think there already are some unit tests that need this.

Reply:
I think we need to move this `assert(pkg_iterator)` under the `err` check otherwise I am getting a failing test:
`test_cr_xml_parse_main_metadata_together_02_invalid_path`